harness/determined/pytorch/_pytorch_context.py

```diff
@@ -57,4 +85,311 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        if self.current_batch_idx is None:
+            raise det.errors.InternalException("Training hasn't started.")
+        return (self.current_batch_idx + 1) % self.hvd_config.aggregation_frequency == 0
+
+    def _backward(
+        self,
+        loss: torch.Tensor,
+        gradient: Optional[torch.Tensor] = None,
+        retain_graph: bool = False,
+        create_graph: bool = False,
+    ) -> None:
+        """backward computes the gradient of current tensor w.r.t. graph leaves.
+
+        The arguments are used in the same way as torch.Tensor.backward().
+        See https://pytorch.org/docs/1.4.0/_modules/torch/tensor.html#Tensor.backward for details.
+
+        Arguments:
+            gradient (Tensor or None): Gradient w.r.t. the
+                tensor. If it is a tensor, it will be automatically converted
+                to a Tensor that does not require grad unless ``create_graph`` is True.
+                None values can be specified for scalar Tensors or ones that
+                don't require grad. If a None value would be acceptable then
+                this argument is optional.
+            retain_graph (bool, optional): If ``False``, the graph used to compute
+                the grads will be freed. Note that in nearly all cases setting
+                this option to True is not needed and often can be worked around
+                in a much more efficient way. Defaults to the value of
+                ``create_graph``.
+            create_graph (bool, optional): If ``True``, graph of the derivative will
+                be constructed, allowing to compute higher order derivative
+                products. Defaults to ``False``.
+        """
+        if self._use_amp:
+            if loss not in self.scaler_mapping:
+                self.scaler_mapping[loss] = self.loss_num
+                self.loss_num += 1
+
+            with apex.amp.scale_loss(
+                loss, self.optimizers, loss_id=self.scaler_mapping[loss]
+            ) as scaled_loss:
+                scaled_loss.backward(
+                    gradient=gradient, retain_graph=retain_graph, create_graph=create_graph
+                )
+
+                if self.hvd_config.use and self._should_communicate_and_update():
+                    # When using horovod, we need to finish communicating gradient
+                    # updates before they are unscaled which happens when we exit
+                    # of this context manager.
+                    for optimizer in self.optimizers:
+                        optimizer.synchronize()
+        else:
+            loss.backward(  # type: ignore
+                gradient=gradient, retain_graph=retain_graph, create_graph=create_graph,
+            )
+
+            # Communication needs to be synchronized so that is completed
+            # before we apply gradient clipping and `step()`.
+            if self.hvd_config.use and self._should_communicate_and_update():
```

Comment:
question: shouldn't this get called as part of `_step_optimizer()`?  It seems like this is going to behave incorrectly in the case that the user calls `_backward()` on more than one loss.

Reply:
Right. Nice catch..

Reply:
Yeah this is going to perform communication for every call to `backwards`: https://github.com/horovod/horovod/blob/9de29ca469a57578b1098d20ac6f4262315be21a/horovod/torch/optimizer.py#L137

The tricky case is with amp, since we need to perform the synchronization before unscaling which happens at `__exit__`.

Reply:
Discussed offline:

Because the library apex we are relying on to support AMP isn’t 100% compatible with horovod, so the implementation of this project might slow down the training if users use AMP, d-train, multiple losses, and multiple optimizers at the same time. The degree of the slowdown depends on the number and sizes of parameters to synchronize. In the future, we will support Torch native AMP, which is not in the scope of this project. We will wait until then to resolve this issue.